basic_system/src/system_functions/modexp/mpnat/arith.rs

```diff
@@ -0,0 +1,97 @@
+// Adopted from https://github.com/aurora-is-near/aurora-engine
+
+use super::u256::U256;
+use zk_ee::system::logger::Logger;
+
+use super::{
+    double::U512,
+    mpnat::{MPNatU256, U256_ZERO},
+};
+use core::{alloc::Allocator, mem::MaybeUninit};
+
+static mut U512_SCRATCH: U512 = U512::zero_const();
+static mut ZERO: MaybeUninit<U256> = MaybeUninit::uninit();
+static mut ONE: MaybeUninit<U256> = MaybeUninit::uninit();
+
+pub fn init() {
+    unsafe { ZERO = MaybeUninit::new(U256::zero()) };
+    unsafe { ONE = MaybeUninit::new(U256::one()) };
+}
+
+/// Computes `(x * y) mod 2^(WORD_BITS*out.len())`.
+/// x and y can't reference RO memory.
+pub unsafe fn big_wrapping_mul<L: Logger>(l: &mut L, x: &[U256], y: &[U256], out: &mut [U256]) {
+    let zero = unsafe { ZERO.assume_init_mut() };
+    let one = unsafe { ONE.assume_init_mut() };
+    let s = out.len();
+    let mut double = unsafe { &mut U512_SCRATCH };
+    let mut c = MaybeUninit::uninit();
+    for i in 0..s {
+        unsafe { zero.write_into_ptr_unchecked(c.as_mut_ptr() as *mut _) };
+        let c = unsafe { c.assume_init_mut() };
+        for j in 0..(s - i) {
+            let x = match x.get(j) {
+                Some(x) => x,
+                None => &zero,
+            };
+            unsafe {
+                shifted_carrying_mul(
+                    l,
+                    &out[i + j],
+                    &x,
+                    y.get(i).unwrap_or(&zero),
+                    &c,
+                    double,
+                    one,
+                )
+            };
+            c.clone_from(double.high());
+            out[i + j].clone_from(double.low());
+        }
+    }
+}
+
+/// Computes `a + xy + c` where any overflow is captured as the "carry",
+/// the second part of the output. The arithmetic in this function is
+/// guaranteed to never overflow because even when all 4 variables are
+/// equal to `Word::MAX` the output is smaller than `DoubleWord::MAX`.
+/// Safety: `x`,`y` can't be placed in RO memory.
+unsafe fn shifted_carrying_mul<L: Logger>(
```

Comment:
Actually that was one of the reasons why I suggested to use raw precompile access - it has built-in carry propagation, so you can do as:
- x * y -> (low, high)
- (low, of1) = low + a
- (low, of2) = low + c
- of = of1 + of2 (carry is at most 2)
- high += of (no carry possible here)

If you prepare scratch space for of -> U256 then it's 1 less precompile call, and hopefully less stack manipulations